q2cli/builtin/tools.py

```diff
@@ -195,7 +196,104 @@ def peek(path):
+@click.option('--cast', required=True, metavar='COLUMN:TYPE', multiple=True,
+              help='Cast parameter for each metadata column that should'
+              ' contain a specified column type (supported types can be'
+              ' found on docs.qiime2.org). The required formatting for this'
+              ' parameter is --cast COLUMN:TYPE, repeated for each column'
+              ' and the associated column type it should be cast to in'
+              ' the resultant output.')
+@click.option('--ignore-extra', is_flag=True,
+              help='Cast parameters that do not correspond to any of the'
+              ' column names within the provided metadata will result in'
+              ' a raised error unless this flag is enabled.')
+@click.option('--error-on-missing', is_flag=True,
+              help='If this flag is enabled, failing to include cast'
+              ' parameters for all columns in the provided metadata will'
+              ' result in an error.')
+@click.option('--output-file', required=True,
+              type=click.Path(exists=False, file_okay=True, dir_okay=False,
+                              writable=True),
+              help='Path to file or directory where the'
+              ' modified metdata should be written to.')
+@click.argument('paths', nargs=-1, required=True, metavar='METADATA...',
+                type=click.Path(exists=True, file_okay=True, dir_okay=False,
+                                readable=True))
+def cast_metadata(paths, cast, output_file, ignore_extra,
+                  error_on_missing):
+    import tempfile
+    import qiime2
+    import numpy as np
+
+    metadata = _merge_metadata(paths)
+
+    cast_dict = {}
+    try:
+        for casting in cast:
+            col, type = casting.split(':')
+            if col in cast_dict:
+                raise click.BadParameter(
+                    message=(f'Column name "{col}" appears in cast more than'
+                             ' once.'),
+                    param_hint='cast')
+            cast_dict[col] = type
+    except Exception as err:
+        header = \
+            ('Could not parse provided cast arguments into unique COLUMN:TYPE'
+             ' pairs. Please make sure all cast flags are of the format --cast'
+             ' COLUMN:TYPE')
+        q2cli.util.exit_with_error(err, header=header)
+
+    types = set(cast_dict.values())
+    if not types.issubset(_COLUMN_TYPES):
+        raise click.BadParameter(
+            message=('Unknown column type provided. Please make sure all'
+                     ' columns included in your cast contain a valid column'
+                     ' type. Valid types: %s' %
+                     (sorted(_COLUMN_TYPES))),
+            param_hint='cast')
+
+    column_names = set(metadata.columns.keys())
+    cast_names = set(cast_dict.keys())
+
+    if not ignore_extra:
+        if not cast_names.issubset(column_names):
+            cast = np.setdiff1d(cast_names, column_names)
```

Comment:
This kind of set difference is actually pretty straightforward to handle in python without pulling in something like `numpy`, you just have to know where to look, which is a challenge!

https://docs.python.org/3.8/library/stdtypes.html#set-types-set-frozenset

example:

```python
x = {1, 2, 3, 4}
y = {3, 4, 5, 6}

set_diff = x - y  # or set_diff = x.difference(y)
assert set_diff == {1, 2}
```